daffodil-runtime1/src/main/scala/org/apache/daffodil/udf/UserDefinedFunctionService.scala

```diff
@@ -0,0 +1,185 @@
+            log(LogLevel.Warning, wstr)
+          }
+          val key = s"$fns:$fname"
+          if (functionProviderLookup.contains(key)) {
+            val estr = s"User Defined Function ignored: ${fcClassName} from provider ${providerClassName}. Duplicate $key found."
+            log(LogLevel.Error, estr)
+          } else {
+            functionProviderLookup += (key -> provider)
+            functionClassesLookup += (key -> fc)
+          }
+        }
+      }
+    }
+  }
+
+  if (functionClassesLookup.isEmpty) {
+    log(LogLevel.Warning, "No user defined functions found. Check that UDF JARs are on classpath and that they are properly registerable by ServiceLoader.")
+    log(LogLevel.Info, s"Current classpath locations:\n${currentClassPath}")
+  }
+
+  def getUserDefinedFunctionClasses = functionClassesLookup.values.toArray
+
+  lazy val allFunctionClasses: String = getUserDefinedFunctionClasses.map {
+    c => s"[${c.getName()} => ${c.getAnnotation(classUserDefinedFunctionInfo)}]"
+  }.mkString("\n")
+
+  /**
+   * Returns an initialized UserDefinedFunction object from the UserDefinedFunctionProvider
+   * based on namespaceURI and function name from the schema
+   *
+   * @return Optional Tuple of initialized UserDefinedFunction, evaluate Method, array of
+   * the evaluate's parameter types and evaluate's return type
+   */
+  def lookupUDFCallingInfo(namespaceURI: String, fname: String): Option[UdfCallingInfo] = {
+    val udfid = s"$namespaceURI:$fname"
+
+    val udfFunctionCallingInfo = InitializedUserDefinedFunctionsCache.getOrElse(udfid, {
+      val provider = functionProviderLookup.get(udfid)
+      val fClass = functionClassesLookup.get(udfid)
+      if (!List(provider, fClass).exists(_.isEmpty)) {
+        val udf = provider.get.lookupInitializedUserDefinedFunction(namespaceURI, fname)
+        val expectedUdfClass = fClass.get
+
+        if (udf != null) {
+          val actualUdfClass = udf.getClass
+          if (actualUdfClass != expectedUdfClass) {
+            log(LogLevel.Error, s"User Defined Function Class Mismatch: $udfid. Expected: $expectedUdfClass Actual: $actualUdfClass")
+            None
+          } else {
+
+            val evaluateMethodInfoTuples = expectedUdfClass.getMethods.collect {
+              case p if p.getName == evaluateMethodName => (p, p.getParameterTypes, p.getReturnType)
+            }
+
+            if (evaluateMethodInfoTuples.isEmpty) {
+              log(LogLevel.Error, s"Missing evaluate method for function provided: $udfid")
+              None
+            } else if (evaluateMethodInfoTuples.length > 1) {
+              log(LogLevel.Error, s"Only one evaluate method allowed per function class: $udfid")
+              None
+            } else if (List(classOf[java.lang.Void], classOf[Unit]).exists(_ == evaluateMethodInfoTuples.head._3)) {
+              log(LogLevel.Error, s"Evaluate method for function provided cannot be void: $udfid")
```

Comment:
Should this mention the return type cannot be void? We also don't do any checks on the parameter types. Should we do that here as well?